### PR TITLE
Port parser to antlr4, compile and link a very very simple example

### PR DESCRIPTION
**Pull request by jdpage**
As requested, I made sure not to PR until it does more than it did.

As of right now, it can sort of compile a.gold (but not link it, since I haven't written the linker yet EDIT: THIS IS A LIE, I WROTE A LINKER).

### CODE REVIEW

**Comment by jdpage** [comment]
Mind you, literally none of the tests pass now (which should be fixable) and currently flake8 is having a fit over the output of antlr4 -- need to figure out a way to disable that. So I understand if you don't want to merge until those things are sorted.

**Comment by woodrowbarlow** [comment]
thank you for breaking stuff into lots of files.

if i work on fixing the broken tests, that would be a good way for me to get familiar with these changes.

please make a list of new parser functionality which didn't exist before. we'll need to write new tests for all of it.

File: jeff65/gold/grammar/Gold.py [hunk 1]
```diff
@@ -0,0 +1,2507 @@
...
+        buf.write("\5\16\b\2\u010a\u010f\5\32\16\2\u010b\u010f\5\22\n\2\u010c")
+        buf.write("\u010f\5\20\t\2\u010d\u010f\5\34\17\2\u010e\u0109\3\2")
+        buf.write("\2\2\u010e\u010a\3\2\2\2\u010e\u010b\3\2\2\2\u010e\u010c")
+        buf.write("\3\2\2\2\u010e\u010d\3\2\2\2\u010f\u0112\3\2\2\2\u0110")
+        buf.write("\u010e\3\2\2\2\u0110\u0111\3\2\2\2\u0111\u0113\3\2\2\2")
+        buf.write("\u0112\u0110\3\2\2\2\u0113\u0114\7\2\2\3\u0114%\3\2\2")
+        buf.write("\2\35\61Z]`bknx\177\u0084\u008b\u009a\u00a1\u00aa\u00ac")
+        buf.write("\u00b9\u00c7\u00cc\u00dd\u00e0\u00e5\u00ec\u00fa\u0104")
+        buf.write("\u0106\u010e\u0110")
+        return buf.getvalue()
```
[the remark this answers is not in the document]
Comment (jdpage) [reply jeff65/gold/grammar/Gold.py]: It's the state machine for the parser in binary form, I think. That's a generated file, so it can neither be commented nor broken up, since changing the source file will just cause it to be overwritten.

I'm not generally a huge fan of checking in generated files, but in this case it seemed more convenient to do so.

**Comment by jdpage** [comment]
Brief tour:

- The lexer still lives in lexer.py, but is now a class instead of a generator method.
- The parser is defined in grammar/Gold.g4, which is used to generate the rest of the files in that directory except __init__.py, which reexports as Parser and a base ParseListener
- the output of the Parser is converted from a concrete syntax tree into an abstract syntax tree of AstNodes by AstBuilder (ast.py)
- The AST is progressively rewritten by a sequence of translation passes, a complete list of which can be found in compiler.py. Translation passes are typically grouped by their domain into the file implementing their data classes. Qualified names are used so it's easy to find them.

As of now, the current capabilities are:

- ALL syntax can be handled by the lexer correctly.
- MOST syntax can be processed by the parser into a concrete syntax tree. Some syntax elements aren't quite correct yet. (Also, if you want to test the CST output there's a good pretty-print method I can show you that produces bracketed expressions--walking the CST is a pain)
- Basically only syntactic elements present in a.gold can be converted into the AST, and only really in the forms that they show up.
- Other compiler passes are dramatically limited, and only handle things which are in a.gold. There's a lot of `def get_random(): return 4` kind of stuff happening. An attempt has been made to at least crash on input we can't handle yet (like assignments bigger than one byte).
- there's no type checker really at all. Types are propagated just enough to implement later passes.
- there's no consistent error-reporting mechanism, though location information is preserved until pretty late in the process. Error reporting usually comes in the form of a generic exception, an assert failure, or a compiler pass making an assumption and causing a language crash.

File: jeff65/gold/grammar/Gold.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (woodrowbarlow) [reply jeff65/gold/grammar/Gold.py]: what are the steps to re-generate it?

it looks like it's ultimately string data as far as python is concerned, so even if we need to have it checked in, we should still be able to move its contents into a datafile and have this function read in the datafile and return its contents as a StringIO object.

oh, wait... unless this entire `Gold.py` file is generated? if so, what's your qualm with making generation of this (and other) files part of the build process and `.gitignore`ing the generated files? i don't really see an advantage to checking it in.

[the remark this answers is not in the document]
Comment (jdpage) [reply jeff65/gold/grammar/Gold.py]: Yeah, literally every file in the grammar subdirectory is generated except for `Gold.g4` and `__init__.py`. So the entirety of `Gold.py` is generated, for example.

I checked them in because in order to generate them, you need to have antlr4 installed. Not only that, but you need to have the version of antlr4 that matches the version of the library we use. Even worse, antlr4 is a Java application, so you need Java... and then of course setting all that up on Windows is a pain, and eventually the Gold.g4 file will be stable enough that it won't need to be regenerated very often.

On the other hand, the generated files are text files, so it's not like git is going to throw a fit from having them around; they'll diff efficiently.

Basically, there are no downsides to having them checked in other than them being a little ugly. The upside is that if you don't have to touch the grammar file, everything you need is already present in the repository.

[the remark this answers is not in the document]
Comment (jdpage) [reply jeff65/gold/grammar/Gold.py]: The steps to regenerate it is basically "get everything installed and run `make` in the checkout directory."

I wouldn't be opposed to replacing Antlr4 with some pure-python generator, or even Lark if we can convince it to play nice. Once the initial AST is constructed, there's no dependency on Antlr4, so replacing it doesn't require rewriting the entire compiler.

**Comment by jdpage** [comment]
Since at this point this branch (a) passes tests (admittedly by disabling most of them), (b) can actually output working binaries, and (c) will be the version that we base future work off of, I'd like to request that we fast-forward this branch onto `master`.

File: jeff65/gold/grammar/Gold.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (woodrowbarlow) [reply jeff65/gold/grammar/Gold.py]: okay, i'm fine with having the generated files in source control. i do, however, want to say that checking in generated files *does* have a downside: developers making changes directly to the generated file, because it seems like a natural thing to do. or the .g4 file and the generated file becoming out of sync. so i guess we just need to watch out for that.

you say the antlr version must be extremely specific... that raises some red flags for me. i've worked on a project that depended on a very specific version of a tool (except they didn't realize it was so version dependent) and the vendor only offered downloads for the latest version. nobody kept a copy of the installer for the version they were using. it was an absolute disaster, the build scripts had to be rewritten, and the entire team upgraded to the latest version of the tool and they stored a copy of the installer in a safe place. obviously that wasn't the best solution, but they were already too dependent on the tool.
you say the ast is generic enough that our dependency on antlr is relatively loose, so that's encouraging. lets just try to keep it that way.
